DoSomething/gambit-conversations
# Change Topic macro [pr]

#### What's this PR do?

Gets the  `changeTopic` macro from #336 working, setting a `conversation.topic` to the Contentful Id `response` reference field of a`defaultTopicTrigger` (either a `textPostConfig`, `textPostConfig`, or `externalPostConfig`entry). This means once we're ready, we can deprecate the `keyword` content type (and related functions) by publishing `defaultTopicTrigger` entries for our current campaigns+topics. Keywords still work as is, handled by the `campaign-keyword` middleware.

* If a campaign keyword is sent, queries `GET /campaigns/:id` endpoint for reply content, using the first entry found in the campaign `topics` array property.

* If a `defaultTopicTrigger` executes a changeTopicMacro, queries `GET /topics/:id` endpoint for reply content, and sets the `conversation.campaignId` based on the topic `campaign` property.

    * Renames the `campaign-current` middleware as `topic-get` in `lib/middleware/messages/member`<p>

* Renames functions like `replies.continueCampaign` and macros like `confirmedCampaign` as `continueTopic`, `confirmedTopic`


* Renames the `campaignMenu` macro as just `menu`, in case we ever want to pitch topics without campaigns, or profile update topics (e.g. #340)


* Renames the `gambit` macro we use for our [catch-all trigger in Rivescript](https://www.rivescript.com/docs/tutorial#catch-all-trigger) as `catchAll`, to improve readability. The `gambit` macro name is a relic from when this app was a MVP named  [`slothie`](https://github.com/DoSomething/slothie)
 and was posting to [`gambit`](https://github.com/DoSomething/gambit).

    * Consolidates the `campaign-closed`, `template-no-campaign`, `parse-ask-signup-response`, `parse-ask-continue-response`, and `campaign-continue` middleware into a `macro-catch-all` middleware. I was a little conflicted about merging them, but each one is simple enough it made it feel more confusing to know which order we check the different conversation states in. Curious if you think this helps or hurts simplification?


#### How should this be reviewed?

* Verify that keyword entries, e.g. `'mirror'`, work as expected, as we can't unpublish them before this is deployed to production.
    * Topic should be set to the first topic found in the `GET /campaigns/:id` response for a campaign

* Verify that `defaultTopicTrigger` entries with topics set work as expected:
    * Run `gambit-campaigns` in Preview mode and this branch of `gambit-conversations` 
    * Verify draft `defaultTopicTrigger` entries change the conversation topic and campaign as expected, e.g. `spit` should change topic to `tv7e98JGXmMM2kskGaUA2` and campaign to 2178.

#### Any background context you want to provide?

* This doesn't update the broadcast or signup message routes to set the topic Id as topic, but they do work. If we send a broadcast this week, it will be setting the rivescript topic to a [hardcoded one](https://github.com/DoSomething/gambit-conversations/blob/2.8.1/brain/topics.rive), so we don't need to worry about setting a contentful Id as a `topic` 

* This needs way more unit tests, but I wanted to open at this point to make it readable.

#### Relevant tickets

Ground work for:

* `POST /v2/messages?origin=afterSignupReminder` endpoint that will check a user's conversation topic and send a reminder template message if it exists for the user's current topic: https://www.pivotaltracker.com/story/show/156174270

* Refactoring the Broadcast content type to reference a topic entry instead of a `campaign` entry: https://www.pivotaltracker.com/story/show/157369418

* A Typeform topic, where a campaign may be optional: https://www.pivotaltracker.com/story/show/157642858

#### Checklist
